content/clause-interleaving.md

```diff
@@ -0,0 +1,180 @@
+~~~
+but that is also invalid because type parameter lists must always come first.
+
+A workaround is to return an intermediate object with an `apply` method, as follows:
+~~~ scala
+class Store:
+  final class StoreGetOrElse[K <: Key](val key: K):
+    def apply[V >: key.Value](default: => V): V = …
+  def getOrElse(key: Key): StoreGetOrElse[key.type] = StoreGetOrElse(key)
+~~~
+This definition provides the expected source API at call site, but it has two issues:
+* It is more complex than expected, forcing a user looking at the API to navigate to the definition of `StoreGetOrElse` to make sense of it.
+* It is inefficient, as an intermediate instance of `StoreGetOrElse` must be created for each call to `getOrElse`.
+* Overloading resolution looks at clauses after the first one, but only in methods, the above is ambiguous with any `def getOrElse(k:Key): ...`, whereas the proposed signature is not ambiguous with for example `def getOrElse(k:Key)[A,B](x: A, y: B)`
+
+Another workaround is to return a polymorphic function, for example:
+~~~scala
+def getOrElse(k:Key): [V >: k.Value] => (default: V) => V = 
+    [V] => (default: V) => ???
+~~~
+While again, this provides the expected API at call site, it also has issues:
+* The behavior is not the same, as `default` has to be a by-value parameter
+* The definition is hard to visually parse, as users are more used to methods (and it is our opinion this should remain so)
+* The definition is cumbersome to write, especially if there are a lot of term parameters
+* Methods containing curried type clauses like `def foo[A][B](x: B)` cannot be represented in this way, as polymorphic methods always have to have a term parameter right after.
+* It is inefficient, as many closures must be created for each call to `getOrElse` (one per term clause to the right of the first non-initial type clause).
+* Same problem as above with overloading
+
+## Proposed solution
+### High-level overview
+
+To solve the above problems, we propose to generalize method signatures so that they can have multiple type parameter lists, interleaved with term parameter lists and using parameter lists.
+
+For the heterogeneous key-value store example, this allows to define `getOrElse` as follows:
+~~~ scala
+def getOrElse(key: Key)[V >: key.Value](default: => V): V = …
+~~~
+It provides the best of all worlds:
+* A convenient API at call site
+* A single point of documentation
+* Efficiency, since the method erases to a single JVM method with signature `getOrElse(Object,Object)Object`
+
+### Specification
+We amend the syntax of def parameter clauses as follows:
+
+~~~
+DefDcl                 ::=  DefSig ‘:’ Type
+DefDef                 ::=  DefSig [‘:’ Type] ‘=’ Expr
+DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
+DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
+DefParamClauseChunk    ::=  [DefTypeParamClause] TermOrUsingParamClause {TermOrUsingParamClause}
```

Comment:
does this mean that this is not valid anymore:
```
  def foo[A](implicit a: A) = ...
```

Reply:
Oh, nicely spotted, my intent was for it to be allowed,
but it appears I made a mistake with the grammar

Since the implementation team (which is also myself) can make changes to the proposal, I believe this can fall under this umbrella
I will think about what's the clearest way to fix the grammar

Reply:
This should fix it:
```diff
  DefDef                 ::=  DefSig [‘:’ Type] ‘=’ Expr
- DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
+ DefSig                 ::=  id [DefParamClauses] [DefTypeParamClause] [DefImplicitClause]
  DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
  DefParamClauseChunk    ::=  [DefTypeParamClause] TermOrUsingParamClause {TermOrUsingParamClause}
```
There is another quirk of this definition however, it allows to parse `def f(a: Int)(b: Int)` to be parsed in two different ways (one vs two `DefParamClauseChunk`), is this an issue ?

Reply:
I didn’t have a look at the complete grammar but what you propose does not look correct, what about the following?

~~~ text
DefDef                 ::=  DefSig [‘:’ Type] ‘=’ Expr
DefSig                 ::=  id [DefParamClauses]
DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
DefParamClauseChunk    ::=  DefTypeParamClause | TermOrUsingParamClause
~~~

Reply:
Alas, this allows two consecutive `type` clauses, or multiple `implicit` clauses.

here's another option:
```
DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
DefParamClauses        ::=  DefParamClauseChunkTypeOpt {DefParamClauseChunkTypeReq}
DefParamClauseChunkTypeOpt    ::=  [DefTypeParamClause] {TermOrUsingParamClause}
DefParamClauseChunkTypeReq    ::=  DefTypeParamClause {TermOrUsingParamClause}
```

Reply:
ah, that doesn't work either... perhaps this:
```
DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
DefParamClauses        ::=  [DefTypeParamClause] {DefParamClauseChunk} [TermOrUsingParamClause]
DefParamClauseChunk    ::=  TermOrUsingParamClause {TermOrUsingParamClause} DefTypeParamClause
```

Reply:
> This should fix it:
> 
> ```diff
>   DefDef                 ::=  DefSig [‘:’ Type] ‘=’ Expr
> - DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
> + DefSig                 ::=  id [DefParamClauses] [DefTypeParamClause] [DefImplicitClause]
>   DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
>   DefParamClauseChunk    ::=  [DefTypeParamClause] TermOrUsingParamClause {TermOrUsingParamClause}
> ```
> 
> There is another quirk of this definition however, it allows to parse `def f(a: Int)(b: Int)` in two different ways (one vs two `DefParamClauseChunk`), is this an issue ?

> I didn’t have a look at the complete grammar but what you propose does not look correct

Could you expand on this, I can't manage to find a failing example with the above syntax

Reply:
I talked about it some more with @julienrf and we came to a very easy and clean solution:
Use the pre-"no type currying" syntax, and add an end of line comment that says "and type clauses cannot be adjacent"

Reply:
```diff
- DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
+ DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]    -- and type clauses cannot be adjacent
- DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
- DefParamClauseChunk    ::=  [DefTypeParamClause] TermOrUsingParamClause {TermOrUsingParamClause}
- TermOrUsingParamClause ::=  DefTermParamClause
-                          |  UsingParamClause
+ DefParamClauses        ::= DefParamClause { DefParamClause }
+ DefParamClause         ::= DefTypeParamClause
+                          | DefTermParamClause
+                          |  UsingParamClause
```
I think this syntax is easier to understand than the formal one

@julienrf How should I go about updating the proposal ?

Reply:
Please open a PR in this repository with your changes.